(Network) poll support

Cthulhu-throwaway opened the pull request:
## Description

This PR adds poll support for platforms that support it through two functions.

#### socket_poll

Low level platform specific function. Same parameters and return values as POSIX's poll.
Only available to platforms defining NETWORK_HAVE_POLL.

#### socket_wait

High level function that only works with a single descriptor. If poll is available, socket_wait will use it, otherwise it falls back to select.
Parameters "rd" and "wr" define which state(s) the function will query/wait. On return, these parameters are set to true/false depending on the state's readiness.
Returns true on success.

Example:
```
bool check_read  = true;
bool check_write = false;
int  timeout     = 5000; /* 5 seconds */

if (socket_wait(fd, &check_read, &check_write, timeout))
{
   if (check_read)
      RARCH_LOG("FD ready to read\n");
   else
      RARCH_LOG("FD not ready to read\n");
}
else
   RARCH_ERR("socket_wait failed")
```

### TODO

Replace socket_select calls with socket_wait and socket_poll calls where appropriate.

### Considerations

There were many problems (non C89 compliant, memory leaks, bugs, etc) within platform-specific code in net_compat.c. This PR includes multiple fixes in there that I cannot test myself.

Also, socket_select was completely broken for PS Vita; socket_select is required by several netplay features.
I've refactored it with what I could research (no docs for Vita's epoll, only function/parameter names and struct/member names); again, I cannot test this myself.

~~And finally, while Vita supports epoll, socket_poll is unavailable due to the lack of documentation (thanks Sony). My refactoring of socket_select also only works with a single descriptor. Maybe someone with more experience developing for this platform can try to solve these limitations (I might try solving them myself eventually).~~

### Broken Builds

I've tried to cover only platforms which have obvious poll support, but I can't guarantee that all of them will build just fine (tested on Windows and Linux only).
It's a very simple matter to blacklist a platform from compiling with poll support, so just notify me if a build happens to fail.

LibretroAdmin commented:
Alright, let me try asking some of the people that develop for some of the homebrew platforms.

@MrHuu Any chance you could test this PR on your end on 3DS and check for any networking issues?

@gblues Would you be able to test this PR on the WiiU and check for any networking issues?

Cthulhu-throwaway commented:
Seems like Vita's epoll is almost the same as Linux's epoll.

I'll be adding proper select and poll support for it and update this PR.

Cthulhu-throwaway commented:
Ok, proper select support added for Vita. It was a pain in the ass because I wanted it to be POSIX compliant (which I believe it is now).

Now to implement poll support for it.

Cthulhu-throwaway commented:
poll support added for Vita and also cleaned and fixed multiple issues within net_socket.h and net_socket.c.

Once this gets merged, I'll start working on switching current calls to select and socket_select to socket_poll and socket_wait where possible.

Feel free to merge it whenever you like.

MrHuu commented:
The 3DS port builds without issues, using this branch.

However, netplay never really worked well on 3DS.
Unable to create a host on the device itself, and only able to connect to an existing host.
Which is also the reason why the netplay menu is disabled by default on the 3DS platform.

Trying to connect to a host does fail currently:

```
[INFO] [Lobby]: Loading core sdmc:/retroarch/cores/fceumm_libretro.cia with current content
[INFO] [Netplay]: Connecting to europe-west1.relay.retroarch.com|55435 (direct)
[WARN] [Netplay] Cannot set netplay port to close-on-exec. It may fail to reopen.
[ERROR] [Netplay] Failed to connect to host 34.140.160.20 on port 55435.
[ERROR] [Netplay] Failed to set up netplay sockets.
[ERROR] [Netplay] Failed to initialize netplay.
```

Otherwise, the ```Online Updater``` functionality seems unaffected and still works as it did before (threaded video driver has to be disabled). But i'm not sure this PR even addresses this logic.

Unfortunately, i'm quite unfamiliar with networking code.
But in any case, this PR shouldn't cause any new issues building for the 3DS platform.

Cthulhu-throwaway commented:
> Trying to connect to a host does fail currently:

Which version did this start happening?

MrHuu commented:
I'll try to bisect, which may take a while. Once i have more info I'll create a separate issue regarding this, as it's not related to this PR.

Cthulhu-throwaway commented:
> I'll try to bisect, which may take a while. Once i have more info I'll create a separate issue regarding this, as it's not related to this PR.

That warning might be a tip for where to start.

The warning happens when a call to fcntl fails. In some platforms (3DS included), fcntl is also used to make the socket's fd non-blocking, which socket_connect_with_timeout does, and if it fails to make the socket non-blocking, socket_connect_with_timeout returns false, which will result in the above error aswell.

Cthulhu-throwaway commented:
@LibretroAdmin any plans to merge this soon? I need this implemented before I can continue working.

LibretroAdmin commented:
@MrHuu You think this is safe to merge as-is?

MrHuu commented:
Yes, issues related to netplay on 3ds are not affected by this.

I've noticed issues with the fcntl and connect calls, forcing both functions to return true allows to connect to external host again. 

This can be safely merged, i'll try and work out the existing issues.

LibretroAdmin commented:
@Cthulhu-throwaway You broke several cores for the Wii now -

https://git.libretro.com/libretro/beetle-pcfx-libretro/-/jobs/2071621

Cthulhu-throwaway commented:
> @Cthulhu-throwaway You broke several cores for the Wii now as a result of these changes -
> 
> https://git.libretro.com/libretro/beetle-pcfx-libretro/-/jobs/2071621
> 
> https://git.libretro.com/libretro/beetle-ngp-libretro/-/jobs/2071643
> 
> We need a fix sooner rather than later

It's a problem with the macro expansion for select. Submitting a PR soon.

LibretroAdmin commented:
WSAPoll breaks backwards compatibility with Windows XP and earlier.

LibretroAdmin commented:
This should deal with the issue I hope -

https://github.com/libretro/RetroArch/commit/d832ce15d6eccdd40953213321142a2fcca91b3b

What I have done, is only selectively define NETWORK_HAVE_POLL for platform builds that target Vista and above. So the regular Mingw-w64 version, MSVC2012, 2013, 2015, 2017, 2019, and UWP versions should all have poll support through WSAPoll. Anything below (VC6/MSVC2003/MSVC2005/MSVC2008/MSVC2010) will resort to select() instead.
